### PR TITLE
Ignition Gazebo support

### PR DESCRIPTION
**Pull request by roni-kreinin**
## Description

- Added Ignition Gazebo support for the Create3.
- Restructured packages to separate Classic and Ignition gazebo.
- Moved center of gravity forwards by 22.8mm on the Create3 body for Ignition.
- Adjusted position of caster wheel in Ignition. This accounts for the different physics behaviours between the DART and ODE physics engines.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

```bash
# Run this command
ros2 launch irobot_create_ignition_bringup create3_ignition.launch.py
```

## Notes
- Bumper contact sensor is not working. This seems to be an issue with `ros_ign_bridge` and spawning the robot from the launch file rather than from the world .sdf file
- It is recommended to install Ignition Gazebo from source. I made some improvements to the Diff Drive plugin parameters which makes driving the robot much better. Installing the binary still works.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: irobot_create_common/irobot_create_description/urdf/create3.urdf.xacro [hunk 1]
```diff
@@ -21,6 +21,13 @@
   <xacro:property name="body_radius"             value="${16.4*cm2m}" />
   <xacro:property name="body_length"             value="${6*cm2m}" />
 
+  <xacro:if value="${'$(arg gazebo)' == 'ignition'}">
+    <xacro:property name="body_cog_x_offset"       value="${2.28*cm2m}" />
+  </xacro:if>
+  <xacro:if value="${'$(arg gazebo)' == 'classic'}">
+    <xacro:property name="body_cog_x_offset"       value="${0*cm2m}" />
```
Comment (alsora) [inline irobot_create_common/irobot_create_description/urdf/create3.urdf.xacro:28]: does moving to DART engine allows to use common parameters?

File: irobot_create_common/irobot_create_description/urdf/create3.urdf.xacro [hunk 2]
```diff
@@ -262,8 +274,20 @@
         <frame_id>odom</frame_id>
         <child_frame_id>base_link</child_frame_id>
         <odom_publish_frequency>50</odom_publish_frequency>
-        <min_acceleration>-1</min_acceleration>
-        <max_acceleration>1</max_acceleration>
+        <!-- Requires ignition gazebo to be installed from source -->
```
Comment (alsora) [inline irobot_create_common/irobot_create_description/urdf/create3.urdf.xacro:277]: What's the purpose of this?
There are parameters only needed by the "master" ignition built from sources?

File: README.md [hunk 3]
```diff
@@ -6,72 +6,4 @@ This is a ROS 2 simulation stack for the [iRobot® Create® 3](https://edu.irobo
...
-colcon build --symlink-install
-source install/local_setup.bash
-```
-
-Then you can run:
-
-```bash
-ros2 launch irobot_create_gazebo_bringup create3_gazebo_aws_small.launch.py
-```
+Both Classic Gazebo and Ignition Gazebo are supported.
```
Comment (alsora) [inline README.md:9]: I would keep the instructions in this page.
 - in the prerequisites we should mention that either classic or ignition is needed
 -  in the examples section we should split it into two sections (one for classic and one for ignition)
 - we should have a section that describes the directory layout of this repo

File: irobot_create_ignition/dependencies.repos [hunk 4]
```diff
@@ -0,0 +1,17 @@
+repositories:
```
Comment (alsora) [inline irobot_create_ignition/dependencies.repos:1]: I think we should have 3 repos files: common, classic and ignition.
This one would basically be the common.
All repos files should be in the root of the repository

File: irobot_create_ignition/ignition_edifice.repos [hunk 5]
```diff
@@ -0,0 +1,65 @@
+repositories:
```
Comment (alsora) [inline irobot_create_ignition/ignition_edifice.repos:1]: why do we need to build ignition from sources ?

File: irobot_create_ignition/irobot_create_ignition_bringup/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,32 @@
+cmake_minimum_required(VERSION 3.8)
```
Comment (alsora) [inline irobot_create_ignition/irobot_create_ignition_bringup/CMakeLists.txt:1]: all packages should be optional.
for example gazebo classic packages should include in the cmake
```
find_package(gazebo_dev)
find_package(gazebo_ros)
if(NOT gazebo_dev_FOUND OR NOT gazebo_ros_FOUND)
  message(WARNING "Gazebo not found. Skipping ${PROJECT_NAME}.")
  ament_package()
  return()
endif()
```

and something similar for ignition packages

File: irobot_create_ignition/irobot_create_ignition_bringup/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
+
+# find dependencies
+find_package(ament_cmake REQUIRED)
+# uncomment the following section in order to fill in
+# further dependencies manually.
+# find_package(<dependency> REQUIRED)
```
Comment (alsora) [inline irobot_create_ignition/irobot_create_ignition_bringup/CMakeLists.txt:12]: remove comments here and at the bottom

File: irobot_create_ignition/irobot_create_ignition_bringup/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+find_package(ros_ign_interfaces REQUIRED)
```
Comment (alsora) [inline irobot_create_ignition/irobot_create_ignition_bringup/CMakeLists.txt:14]: why is this needed?

File: irobot_create_common/irobot_create_description/urdf/create3.urdf.xacro (hunk 1 above)
Comment (roni-kreinin) [reply irobot_create_common/irobot_create_description/urdf/create3.urdf.xacro:28]: It should. In ODE this currently makes the robot tilt forwards so I have left it as is for now.

File: irobot_create_common/irobot_create_description/urdf/create3.urdf.xacro (hunk 2 above)
Comment (roni-kreinin) [reply irobot_create_common/irobot_create_description/urdf/create3.urdf.xacro:277]: I made a PR for ign-gazebo which separated the min/max parameters into angular and linear. See https://github.com/ignitionrobotics/ign-gazebo/commit/fe8c1edb3af33b809b6b5ac22ee76308f17a7ba7
This is currently only available if built from sources and I am not sure if they will have a new binary release any time soon. The current binary only has parameters such as `min_acceleration` which apply to both linear and angular acceleration, which is not great for our use case. I have left these in to be compatible with the binary installation should someone choose to use that, but the new `min_linear_acceleration`, `min_angular_acceleration` parameters overwrite these in the source version. They also do not affect the binary version.

File: irobot_create_ignition/ignition_edifice.repos (hunk 5 above)
Comment (roni-kreinin) [reply irobot_create_ignition/ignition_edifice.repos:1]: For the updated DiffDrive plugin. I am also hoping they will soon fix the contact sensor in which case the source install will be updated first for that.

File: irobot_create_ignition/irobot_create_ignition_bringup/CMakeLists.txt (hunk 8 above)
Comment (roni-kreinin) [reply irobot_create_ignition/irobot_create_ignition_bringup/CMakeLists.txt:14]: This contains custom messages for the `ros_ign_bridge`, including the message used for the contact sensor.

**Comment by alsora** [comment]
@roni-kreinin can you resolve the conflicts with the main branch?

**Comment by roni-kreinin** [comment]
Done.

**Comment by alsora** [comment]
Thanks, we are getting close, there are only a few linter errors to fix.

**Review by alsora — APPROVED** [review]

**Review by justinIRBT — APPROVED** [review]
